### PR TITLE
Github plugin visibility change

### PR DESCRIPTION
**Pull request by liu-ziyang**
# Description
Addressed discrepancy between docs and code in plugin visibility to users, update the documentation that we will be using GitHub topics to search for plugins that are not on PyPI, also added a description that this may be deprecated in the future in favor of the PyPI classifier

Added a toggle that by default hide plugins only from Github in favor of PyPI releases, user can unhide the plugins from Github if needed

## Type of change
- New feature (non-breaking change which adds functionality)
- Documentation

# References
some of the backgrounds are discussed in https://github.com/napari/napari-plugin-engine/issues/16

# How has this been tested?
Manually tested the interface

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Carreau** [comment]
Failure seem to be a problem with the coverage package that crashed when combining report.
Someone with rights could restart CI.

**Review by Carreau — APPROVED** [review]

**Comment by sofroniewn** [comment]
> Failure seem to be a problem with the coverage package that crashed when combining report.
Someone with rights could restart CI.

I think right now all/ most of our windows tests are failing right now, independent of this PR - I have tried rerunning, but we should also see if there's anything we can do to fix the current windows CI fail

**Comment by tlambert03** [comment]
> > Failure seem to be a problem with the coverage package that crashed when combining report.
> > Someone with rights could restart CI.
> 
> I think right now all/ most of our windows tests are failing right now, independent of this PR - I have tried rerunning, but we should also see if there's anything we can do to fix the current windows CI fail

i took a crack at that this weekend and couldn't repeat it locally on a windows box unfortunately :/  annoying.  need to take another look soo

**Comment by sofroniewn** [comment]
Huh, all tests are passing! The CI gods are looking favorably upon us right now!

@tlambert03 anything else you'd like to see from this PR before merge/ any refinement to the language in the docs. I worry "we could deprecate github repo matching in the future" might be a little unsettling for people to read as phrased.

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [inline docs/source/plugins/for_plugin_developers.rst]: I would leave this whole bit out. This was already sort of a hidden "workaround" feature... so I'm not sure that we need to _add_ language to point people to the fact that they can do this, just to then say "but actually, don't do this... it may not work soon".  imo, simply mentioning the github topic is sufficient.

File: napari/_qt/dialogs/qt_plugin_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/qt_plugin_dialog.py:406]: being a boolean attribute, this feels more like a QCheckBox than a push button no?  You could leave out the `self._search_github` attribute and the `_toggle_search_github` methods entirely, and just have `create_worker(... search_github=self.search_github_checkbox.isChecked() )

File: docs/source/plugins/for_plugin_developers.rst
Comment (liu-ziyang) [reply docs/source/plugins/for_plugin_developers.rst]: I respectfully disagree, the documentation should explain exactly what the code does, imaging someone else come to visit this code and try to improve it, they would also be confused (same for myself) as to why there is code related to github topics, and logic that checks the setup.py. Given that this loophole exists for backward compatibility, we should mention so for clarity

File: napari/_qt/dialogs/qt_plugin_dialog.py
Comment (liu-ziyang) [reply napari/_qt/dialogs/qt_plugin_dialog.py:406]: yeah, I was originally having it as a checkbox, but then I tried to use the push button to make the UI looking more consistent, checkbox looks kinda ugly in the exsiting setup

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: > why there is code related to github topics, and logic that checks the setup.py. 

I would fully support that comment in the code... I just don't think it belongs on our website documentation

Comment (sofroniewn) [reply docs/source/plugins/for_plugin_developers.rst]: I take your point @ziyangczi about the importance being clear, but I think the proposal from @tlambert03 to add this as a comment in the code rather than on the main website is a good compromise for now.

Simplicity on the main website is also a virtue - most of the people coming to those docs will be interested in just the basic, recommended way to do something. An alternative, advanced option that we might deprecate, is something that we should throughly document in the code instead.

We can also revisit this in the next few weeks as we get more experience with both discovery mechanisms. At that point we could make some more final decisions and update the public website docs accordingly.

Can you make move the doc string to a code comment @ziyangczi? It would be much appreciated

Comment (liu-ziyang) [reply docs/source/plugins/for_plugin_developers.rst]: fair enough, I will move this to code comment instead

**Comment by liu-ziyang** [comment]
may I get an approval to merge in this? github would not allow me to merge, unless anything else we want to address?

File: napari/plugins/pypi.py
Comment (sofroniewn) [inline napari/plugins/pypi.py:109]: This doc string is no longer accurate - we now look for the topic not the readme

File: napari/_qt/dialogs/qt_plugin_dialog.py
Comment (sofroniewn) [reply napari/_qt/dialogs/qt_plugin_dialog.py:406]: I find it strange not to have a checkbox too, maybe the checkbox can go on the same row as the `Available Plugins` title but aligned to the right hand margin (so just above the blue `install` button) and be fairly discreet
Screen Shot 2021-02-04 at 7 54 58 PM

**Review by sofroniewn — COMMENTED** [review]
I've left a couple comments, one on the checkbox/ one on a doc string that is now out of date.

I'm not convinced that exposing this choice in the GUI to users right now is critical. When I tried it today there were only 3 plugins that were found only via GitHub and I think it would be pretty easy to submit PRs to all 3 projects adding the PyPI classifier.

Has there been any discussion yet of just leaving this as `_search_github = False` or dropping the GitHub search methods entirely?

It seems like now we require all our plugins to be on PyPI anyway as we run `ensure_published_at_pypi`. I know pip can be used to install repos directly from github, without requiring them to be on PyPI, and that was something we discussed before but it doesn't look like we're doing that now.

We might just want to drop the github search until we have a greater need for it. I could imagine a future where we search github more directly, or something conda-forge related etc, but I don't think we're there yet.

Note that the website docs telling people 

```
If you are using Github, add the `"napari-plugin" topic
<https://github.com/topics/napari-plugin>`_ to your repo so other developers can
see your work.
```

is still relevant as the GitHub topic is a convenient way to allow other developers on GitHub to find your plugin through the GitHub search functionality, but using it to find plugins really was a work around until we got the PyPI classifier approved (which took a few months).

What do you think @tlambert03?

**Comment by liu-ziyang** [comment]
I would love to remove the github reference all together! I did not propose this because I don't know what impact it would have both on the current installations and future proofness, the rationale is more on the future proofness if someone would not be releasing to pypi at all

**Comment by tlambert03** [comment]
I'd be totally ok with removing github visibility altogether!

**Comment by sofroniewn** [comment]
Sounds like all three of us are in agreement that removing the github visibility entirely is now fine. 

Maybe just remove the Qt changes and leave the `_search_github=False` in this PR @ziyangczi. We can make a followup PR to remove the code entirely later if we want (I'm also ok dropping the code entirely now too, it's not too hard to add back in).

**Comment by liu-ziyang** [comment]
gonna drop the github code instead

**Pull request closed without merging**
